enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,279 @@
+
+There is no way to provide [MachineSpecific Configuration in OpenShift](https://github.com/openshift/machine-config-operator/issues/1720) so I am
+forced to use the UPI flow which is less automated and more prone to errors.
+
+### Risks and Mitigations
+
+In some existing workflows, kubernetes-nmstate is used to do network configuration on day-2. Using a different interface for day-1 introduces the potential for mismatches and configuration errors when making day-2 changes. It should be possible to mitigate this by generating the NetworkManager keyfiles using the nmstatecli, which will allow the same nmstate config to be used directly in the day-2 configuration.
+
+An added benefit of writing nmstate configurations instead of NetworkManager
+keyfiles directly is that if/when a day-1 nmstate-based interface becomes
+available the config files can be directly inserted into the new interface.
+However, this is not a solution because it still requires manually copying
+configuration to multiple locations which introduces the potential for errors.
+
+## Design Details
+
+In the IPI flow day-1 network configuration is required in 2 different cases:
+
+* Deployment of the controlplane hosts via terraform, using input provided to openshift-install
+* Deployment of compute/worker hosts (during initial deployment and scale-out), via Machine API providers for each platform
+
+In the sections below we will describe the user-facing API that contains network configuration, and the proposed integration for each of these cases.
+
+### User-facing API
+
+RHCOS already provides a mechanism to specify NetworkManager keyfiles during deployment of a new node. We need to expose that functionality during the IPI install process. There are a couple of options on how to do that:
+
+* A new section in install-config.
+* A secret that contains base64-encoded content for the keyfiles.
+
+These are not mutually exclusive. If we implement the install-config option, we will still need to persist the configuration in a secret so it can be used for day-2. In fact, the initial implementation should likely be based on secrets, and then install-config integration can be added on top of that to improve the user experience. This way the install-config work won't block implementation of the feature.
+
+The data provided by the user will need to have the following structure:
+```yaml
+[hostname]:
+  [interface name]:
+    [contents of NetworkManager key file for interface]
+  [second interface name]:
+    [contents of NetworkManager key file for second interface]
+  etc...
+```
+
+For example:
+```yaml
+openshift-master-0:
+  eth0:
+    [connection]
+    id=eth0
+    uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+    type=ethernet
+    etc...
+openshift-master-1:
+  eth0:
+    [connection]
+    id=eth0
+    etc...
+```
+
+In install-config this would look like:
+```yaml
+platform:
+  baremetal:
+    hosts:
+      - name: openshift-master-0
+        networkConfig:
```

Comment:
I've been thinking about the implementation. Whilst this looks logically neat, it would be hard to type into golang. My suggestion would be to use the following format for the install-config. Let me know what you think:

```
platform:
  baremetal:
    hosts:
      - name: openshift-master-0
        networkConfig:
          - interface: eth0
            data: |
              [connection]
              id=eth0
               ...
          - interface: eth1
            data: |
              [connection]
              id=eth1
               ...


```